File: 2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: ListNode) -> int:
        slow = head
        fast = head
        while fast != None and fast.next != None:
            slow = slow.next
            fast = fast.next.next
        prev = None
        curr = slow
        while curr != None:
            new_node = curr.next
            curr.next = prev
            prev = curr
            curr = new_node
        second_half = prev
        first_half = head
        max_sub = 0
        while second_half != None:
            current_sum = first_half.val + second_half.val
            max_sub = max(max_sub , current_sum)
            first_half = first_half.next
            second_half = second_half.next
        return max_sub

File: 2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* reverse(struct ListNode* head){
    struct ListNode* prev = NULL;
    struct ListNode* curr = head;
    while(curr != NULL){
        struct ListNode* new_node = curr->next;
        curr->next = prev;
        prev = curr;
        curr = new_node;
    }
    return prev;
}

int pairSum(struct ListNode* head) {
    struct ListNode* slow = head;
    struct ListNode* fast = head;
    while(fast != NULL && fast->next != NULL){
        slow = slow->next;
        fast = fast->next->next;
    }

    struct ListNode* firstHalf = head;
    struct ListNode* secondHalf = reverse(slow);
    int maxSub = 0;

    while(secondHalf != NULL){
        int current_sum = firstHalf->val + secondHalf->val;
        if(current_sum > maxSub){
            maxSub = current_sum;
        }
        firstHalf = firstHalf->next;
        secondHalf = secondHalf->next;
    }
    return maxSub;
}